library/CodeMonkeysRu/GCM/Sender.php

```diff
@@ -94,41 +94,64 @@ public function send(Message $message)
         curl_setopt($ch, CURLOPT_POST, true);
         curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
         curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
-
+        
+        curl_setopt($ch, CURLOPT_HEADER, 1); // return HTTP headers with response
+        
+        
         if ($this->caInfoPath !== false) {
             curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, true);
             curl_setopt($ch, CURLOPT_CAINFO, $this->caInfoPath);
         } else {
             curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
         }
-
+        
+        
         curl_setopt($ch, CURLOPT_POSTFIELDS, $data);
 
-        $resultBody = curl_exec($ch);
+        $resp = curl_exec($ch);
+ 
+        if ($resp === FALSE) {
+            throw new Exception('cURL error: '. curl_error($ch), Exception::CURL_ERROR);
+         
+        }
+        
+        list($responseHeaders, $resultBody) = explode("\r\n\r\n", $resp, 2);
+        // $headers now has a string of the HTTP headers
+        // $resultBody is the body of the HTTP response
+
+        $responseHeaders = explode("\n", $responseHeaders);
+        
         $resultHttpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
 
         curl_close($ch);
-
-        switch ($resultHttpCode) {
-            case "200":
-                //All fine. Continue response processing.
-                break;
-
-            case "400":
-                throw new Exception('Malformed request. '.$resultBody, Exception::MALFORMED_REQUEST);
-                break;
-
-            case "401":
-                throw new Exception('Authentication Error. '.$resultBody, Exception::AUTHENTICATION_ERROR);
-                break;
-
-            default:
-                //TODO: Retry-after
-                throw new Exception("Unknown error. ".$resultBody, Exception::UNKNOWN_ERROR);
-                break;
+        
+        if ($resultHttpCode == "200") {
```

Comment:
I agree. I'll put it back to your version. Now that response headers are included in the Response object the programmer can do the more fine-grained analysis himself if he wants.